tools/prologs-epilogs/gpu-health

```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+#
+# Check that all GPUs are healthy via DCGM and check for ECC errors
+#
+if [ $NUMGPUS -gt 0 ]; then
+    echo "Execute DCGM health check and ECC error check for GPUs"
+    GPULIST=`nvidia-smi --query-gpu=index --format=csv,noheader | tr '\n' ',' | sed 's/,$//'`
+    rm /tmp/dcgm.out 2> /dev/null
+    rm /tmp/ecc_errors.out 2> /dev/null
+
+    # Run DCGM checks
+    nv-hostengine
+    dcgmi group -c gpuinfo
+    dcgmi group -g 1 -a $GPULIST
+    dcgmi diag -g 1 -r 1 1> /tmp/dcgm.out
+    dcgmi group -d 1
+    nv-hostengine -t
+
+    # Check for DCGM failures
+    grep -i fail /tmp/dcgm.out > /dev/null
+    DCGM_FAILED=$?
+
+    # Check for ECC errors (double-bit errors for H100)
+    nvidia-smi --query-gpu=index,ecc.errors.double_bit.volatile.total --format=csv,noheader > /tmp/ecc_errors.out
+    ECC_ERRORS=$(awk -F', ' '{sum += $2} END {print sum}' /tmp/ecc_errors.out)
+
+    # Check GPU utilization to ensure GPUs are accessible
+    GPU_UTIL=$(nvidia-smi --query-gpu=utilization.gpu --format=csv,noheader,nounits | awk '{sum += $1} END {print sum/NR}')
+
+    # Check for NVLink errors
+    NVLINK_ERRORS=$(nvidia-smi nvlink -sc 0bz -i 0 | grep -i "Error Count" | awk '{sum += $3} END {print sum}')
+
+    if [ $DCGM_FAILED -eq 0 ] || [ $ECC_ERRORS -gt 0 ] || [ $(echo "$GPU_UTIL < 1" | bc) -eq 1 ] || [ $NVLINK_ERRORS -gt 0 ]; then
+        REASON="H100 GPU issues detected: "
+        [ $DCGM_FAILED -eq 0 ] && REASON+="DCGM test failed, "
+        [ $ECC_ERRORS -gt 0 ] && REASON+="ECC errors found ($ECC_ERRORS double-bit errors), "
+        [ $(echo "$GPU_UTIL < 1" | bc) -eq 1 ] && REASON+="GPUs showing zero utilization, "
+        [ $NVLINK_ERRORS -gt 0 ] && REASON+="NVLink errors detected ($NVLINK_ERRORS errors), "
+        REASON+="see /tmp/dcgm.out and /tmp/ecc_errors.out"
+
+        scontrol update nodename=$HOSTNAME state=drain reason="$REASON"
```

Comment:
I believe this line is unnecessary as it is already the default behavior:

https://slurm.schedmd.com/prolog_epilog.html#failure_handling

It is also explicitly recommend that `scontrol` commands not be present in prologs/epilogs (see link above):

> Prolog and Epilog scripts should be designed to be as short as possible and should not call Slurm commands (e.g. squeue, scontrol, sacctmgr, etc). Long running scripts can cause scheduling problems when jobs take a long time to start or finish. Slurm commands in these scripts can potentially lead to performance issues and should not be used.

Reply:
if she doesn't do this, is there any other way to set the reason for the drain state?